File: sorting/java/HeapSort.java
public class HeapSort {

    private static void heapify(int[] arr, int n, int i) {
        int largest = i;
        int l = 2 * i + 1;
        int r = 2 * i + 2;

        if (l < n && arr[l] > arr[largest]) {
            largest = l;
        }
        if (r < n && arr[r] > arr[largest]) {
            largest = r;
        }
        if (largest != i) {
            int temp = arr[i];
            arr[i] = arr[largest];
            arr[largest] = temp;
            heapify(arr, n, largest);
        }
    }

    private static void heapSort(int[] arr) {
        int n = arr.length;
        for (int i = n / 2 - 1; i >= 0; i--) {
            heapify(arr, n, i);
        }
        for (int i = n - 1; i > 0; i--) {
            int temp = arr[0];
            arr[0] = arr[i];
            arr[i] = temp;
            heapify(arr, i, 0);
        }
    }

    public static void sort(int[] arr) {
        if (arr == null || arr.length <= 1) {
            return;
        }
        heapSort(arr);
    }

    public static void main(String[] args) {
        int[] arr = { 9, 4, 3, 8, 10, 2, 5 };
        sort(arr);
        for (int v : arr) {
            System.out.print(v + " ");
        }
        System.out.println();
    }
}

File: sorting/python/heap_sort.py
"""
Heap Sort – heapify iterativo para evitar RecursionError con 1 M elementos.
Complejidad: O(n log n) tiempo, O(1) espacio extra.
"""


def _heapify_iterative(arr, n, i):
    """Versión iterativa de heapify: sin riesgo de stack overflow."""
    while True:
        largest = i
        l = 2 * i + 1
        r = 2 * i + 2
        if l < n and arr[l] > arr[largest]:
            largest = l
        if r < n and arr[r] > arr[largest]:
            largest = r
        if largest == i:
            break
        arr[i], arr[largest] = arr[largest], arr[i]
        i = largest


def heapSort(arr):
    n = len(arr)
    # Build max-heap
    for i in range(n // 2 - 1, -1, -1):
        _heapify_iterative(arr, n, i)
    # Extract elements one by one
    for i in range(n - 1, 0, -1):
        arr[0], arr[i] = arr[i], arr[0]
        _heapify_iterative(arr, i, 0)


# Driver
if __name__ == "__main__":
    arr = [9, 4, 3, 8, 10, 2, 5]
    heapSort(arr)
    print("Sorted:", arr)